src/absil/ilwrite.fs

```diff
@@ -4462,105 +4474,146 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
+            writeInt32AsUInt16 os2 idd.iddMinorVersion
+            writeInt32 os2 idd.iddType
+            writeInt32 os2 idd.iddData.Length  // IMAGE_DEBUG_DIRECTORY.SizeOfData 
+            writeInt32 os2 debugDataChunk.addr  // IMAGE_DEBUG_DIRECTORY.AddressOfRawData 
+            writeInt32 os2 (textV2P debugDataChunk.addr)// IMAGE_DEBUG_DIRECTORY.PointerToRawData 
+
+            (* dprintf "idd.iddCharacteristics = %ld\n" idd.iddCharacteristics
+            dprintf "iddMajorVersion = %ld\n" idd.iddMajorVersion
+            dprintf "iddMinorVersion = %ld\n" idd.iddMinorVersion
+            dprintf "iddType = %ld\n" idd.iddType
+            dprintf "iddData = (%A) = %s\n" idd.iddData (System.Text.Encoding.UTF8.GetString idd.iddData) *)
                   
-                // write the debug raw data as given us by the PDB writer 
-                os2.BaseStream.Seek (int64 (textV2P debugDataChunk.addr), SeekOrigin.Begin) |> ignore
-                if debugDataChunk.size < idd.iddData.Length then 
-                    failwith "Debug data area is not big enough.  Debug info may not be usable"
-                writeBytes os2 idd.iddData
-                os2.Close()
-            with e -> 
-                failwith ("Error while writing debug directory entry: "+e.Message)
-                (try os2.Close(); FileSystem.FileDelete outfile with _ -> ()) 
-                reraise()
+            // write the debug raw data as given us by the PDB writer 
+            os2.BaseStream.Seek (int64 (textV2P debugDataChunk.addr), SeekOrigin.Begin) |> ignore
+            if debugDataChunk.size < idd.iddData.Length then 
+                failwith "Debug data area is not big enough.  Debug info may not be usable"
+            writeBytes os2 idd.iddData
         with e -> 
-            reraise()
-            
-    end
-    reportTime showTimes "Finalize PDB"
+            failwith ("Error while writing debug directory entry: "+e.Message)
 
-    /// Sign the binary.  No further changes to binary allowed past this point! 
-    match signer with 
-    | None -> ()
-    | Some s -> 
-        try 
-            s.SignFile outfile
-            s.Close() 
+        reportTime showTimes "Finalize PDB"
+
+    /// Sign the binary.  No further changes to binary allowed past this point!
+    let signTo filePath (s: ILStrongNameSigner) =
+        try
+            reportTime showTimes "Signing Image"
+            use closeS = { new System.IDisposable with member x.Dispose() = try s.Close() with _ -> () }
+            s.SignFile filePath
         with e -> 
             failwith ("Warning: A call to StrongNameSignatureGeneration failed ("+e.Message+")")
-            (try s.Close() with _ -> ())
-            (try FileSystem.FileDelete outfile with _ -> ()) 
-            ()
-
-    reportTime showTimes "Signing Image"
-    //Finished writing and signing the binary and debug info...
+        reportTime showTimes "Signed Image"
+
+    let serializeToFile extension f emitter =
```

Comment:
This looks independent of context (no free variables) - please pull it out to be a top level function